### ISSUE TITLE
Refreshing the page with esbuild doesn't wire up components

### ISSUE THREAD

**Issue by joemasilotti** [hotwired/strada-ios#20]
Refreshing a page hitting a Rails server with importmap correctly wires up the Strada components. If the Rails backend is instead using esbuild the Strada component never fires.

* [esbuild repo](https://github.com/joemasilotti/strada-bug-esbuild)
* [importmap repo](https://github.com/joemasilotti/strada-bug-importmap)
* [turbo-ios demo PR](https://github.com/hotwired/turbo-ios/pull/152)

**Comment by mrtnin** [comment]
I'm using esbuild on the project I'm working on, and I don't see the same issue. Main difference seems to be:
- We import and register js controllers manually
- We use a custom esbuild config

**Comment by joemasilotti** [comment]
@mrtnin, is it possible to comment out pieces of the custom esbuild config to see if you can reproduce the issue?

**Comment by mrtnin** [comment]
I may have spoken too soon @joemasilotti. I see the same issue on my end. When I tested yesterday I was testing on a component that responds to triggered events.

```js
export default class extends BridgeComponent {
  static component = "sign-out"
  static targets = []

  signOut(event) {
    event.preventDefault()
    event.stopImmediatePropagation()
    this.send("signOut", { }, () => {
      console.log("I think I logged out?")
    })
  }
}
```

```html
<a data-controller="bridge--sign-out" data-action="bridge--sign-out#signOut">Sign out</a>
```
The message is sent even after refresh. And it's received by the BridgeComponent on iOS.

When adding a breakpoint in `connect()` and running `document.documentElement.dataset` in console yields:
```js
{bridgePlatform: "ios", bridgeComponents: ""}
```

Maybe it is as simple as script load order?

**Comment by joemasilotti** [comment]
Excellent! Well, not that there's a bug but that we could reproduce it.

**Comment by alhajrahmoun** [comment]
I experienced something similar to this yesterday.

I am using esbuild in my app and once I refresh the page (by pulling down), the button does what it's supposed to do on iOS but the `reply` part doesn't work.

**Comment by noahsettersten** [comment]
We've encountered something similar as well (Turbo Native app using Strada; Rails server's JS bundled with ESBuild). After pulling down to refresh the page in the iOS app, sometimes the Javascript callback for our Strada component isn't called anymore. I added a few breakpoints to the Strada Swift package and a few console logs to the `@hotwired/strada` NPM package and it appears to be a similar issue as @mrtnin: the list of bridge components is empty (at least at first) and the Strada Javascript skips the callback because `this.supportsComponent(component)` returns false.

As a fix for this, I've used `patch-package` to comment out that `supportsComponent` check and it appears to be working successfully now. Perhaps there is a timing issue where the Strada callbacks are set up before the list of bridge components are set on the document element?

**Comment by johnloringpollard** [comment]
We too are experiencing this issue. What is the best solution? I don't want to comment out the check for `supportsComponent` and I don't love adding a timeout to fix the race condition. One solution appears to be changing loading the js script for registering the stimulus components from `defer: true` to `async: true` but I'm nervous that might run into other loading issues.

**Comment by dginsburg** [comment]
I submitted a PR to fix this. https://github.com/hotwired/strada-ios/pull/27

**Comment by joemasilotti** [comment]
Closed via #27.
